shared-bindings/uvc/__init__.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2024 Jeff Epler for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/uvc/__init__.h"
+#include "shared-bindings/uvc/UVCFramebuffer.h"
+#include "shared-module/uvc/__init__.h"
+
+//| """Allows streaming bitmaps to a host computer via USB UVC
+//|
+//| This interface is experimental and may change without notice even in stable
+//| versions of CircuitPython."""
+//|
+//| bitmap: displayio.Bitmap
+//| """The content of this bitmap image is streamed to the host computer. To change the
+//| displayed image, modify the bitmap contents.
+//|
+//| The bitmap content is in RGB565_SWAPPED format. It is converted to "YUYV" format before
+//| transmission to the host computer. The "YUYV" format stores a "Y" (luminance) sample
+//| for every pixel, but a chromonance (UV) value only for each pair of pixels; due to this,
+//| the color of adjacent pixels cannot be fully controlled. For instance, there is no way
+//| for the first pixel to be red and the second pixel to be green. However, they can both
+//| be different shades of red.
+//| """
+//|
+
+//| def enable_framebuffer(width: int, height: int) -> None:
+//|     """Enable the UVC gadget, setting the given width & height
+//|
+//|     During ``boot.py``, this function may be used to enable the UVC gadget.
+//|
+//|     Width is rounded up to a multiple of 2.
```

Comment:
Because of how YUYV works, there are always a multiple of 2 horizontal pixels. This only affects oddball cases: you simply can't allocate a framebuffer that is (say) 95x96 pixels in size when working with the YUYV format.

Reply:
Ah, ok. I had originally read this as power of two.